pytorch_lightning/strategies/single_device.py

```diff
@@ -20,20 +22,21 @@
 from pytorch_lightning.plugins.precision import PrecisionPlugin
 from pytorch_lightning.strategies.strategy import Strategy
 from pytorch_lightning.utilities import _XLA_AVAILABLE
+from pytorch_lightning.utilities.types import _DEVICE
 
 
 class SingleDeviceStrategy(Strategy):
     """Strategy that handles communication on a single device."""
 
     def __init__(
         self,
-        device: torch.device,
-        accelerator: Optional["pl.accelerators.accelerator.Accelerator"] = None,
-        checkpoint_io: Optional[CheckpointIO] = None,
-        precision_plugin: Optional[PrecisionPlugin] = None,
+        device: _DEVICE,
```

Comment:
Now this supports all device types.

Reply:
yes but this is now very limiting, because subclasses can't restrict the type because it is a union of 3 types now. mypy does not allow it.

Reply:
I agree, I think we should omit raw strings and ints here. So far we said that this is not an API each user will use and therefore we can expect people to stick to the API of `torch.device` and I'd advocate to keep it that way.

Reply:
> mypy does not allow to choose only a subset.

@awaelchli can you illustrate the problem exactly? this works:

https://mypy-play.net/?mypy=latest&python=3.10&gist=3491fba8a2c0c70f4ecaab22a2da919a

> So far we said that this is not an API each user will use and therefore we can expect people to stick to the API of torch.device and I'd advocate to keep it that way.

Since the goal is for all this to become stable soon (hence fully public), I'd say it is also expected that we minimize boilerplate on these APIs

Reply:
I illustrate:

class SingleDeviceStrategy:

    def __init__(self, device: Union[int, str, device]):
        pass

class LimitedSingleDeviceStrategy:
    def __init__(self, device: device):
        pass

Now mypy will complain that the super type does not match the type of this class.

Reply:
I was a bit mislead by a related issue with *args **kwargs not working that way, but the above sample actually passes mypy checks.